### ISSUE TITLE
Make it possible to invoke the compiler directly

### ISSUE THREAD

**Issue by danmoseley** [dotnet/sdk#8742]
Not sure whether this belongs in the SDK repo instead.

In the desktop world, most people build with MSBuild on the command line or in VS. But for quick experimentation, learning c#, or custom scripts, one can invoke csc.exe directly. This has the advantage of being very fast, and not needing a project file. In that situation csc adds some default references so you simply run `csc Program.cs`. It is implicit that you are getting no help resolving references, so this is not a good way to make shipping software, unless you really want something custom and know what you're doing. When you are done experimenting, you should make a project file. This scenario has worked well since .NET Framework v1.

Can we support this scenario for .NET Core? This would be a supported shortcut for `dotnet C:\Program Files\dotnet\sdk\...version..\Roslyn\csc` that would also add some basic default references.

cc @ViktorHofer @stephentoub

**Comment by vhnatyk** [comment]
yes it wouldn't hurt, it would not at all so - after two years still have to revive python for single file cross-platform scripting instead of beloved C# 🤦‍♂️

**Comment by jaredpar** [comment]
My personal goal is to make it possible to use `dotnet build` and `dotnet run` for .cs files without needing an explicit project file. That is very different though than allowing direct invocation of the compiler in the form of say `dotnet csc hello-world.cs`. 

The reason for using `dotnet build / run` vs `dotnet csc` for a single file application is because at a fundamental level building and running for .NET Framework and .NET Core are very different problem spaces. In .NET Framework there are the following advantages:

1. While reference assemblies exist, there is really no meaningful difference in compiling against reference vs. implementation assemblies. 
1. Because of (1) above there is no need to every "find the reference assemblies". When you run `csc example.cs` on .NET Desktop we can find all of the default references by using the default csc.rsp file and resolving all of the assemblies in `typeof(object).Assembly.Loctation`
1. There really isn't a deployment step to speak of which means you can easily `csc /t:exe example.cs` and then immediately just run `example.exe`.

That is why for .NET Framework using `csc` as the single tool for building is very approachable. For .NET Core though it's really not that generally useful for customers because they need to solve all of the above problems to successfully use `csc.`. The logic for those steps are inside `dotnet build / run` hence that is what I envision us leveraging for the single *.cs file scenarios.

That being said: I would like to make `dotnet csc` as usable on .NET Core as it is on .NET Framework for the act of **compiling**. If for nothing else than because it's how the Roslyn team does a lot of it's own work :wink:. But I don't think it will be a tool that a ton of customers end up finding that useful.

**Comment by danmosemsft** [comment]
An implicit project  file sounds like a great idea. Is that idea tracked anywhere?

**Comment by jaredpar** [comment]
Not in an official proposal yet. Still working on a draft of this. It's been discussed in a few forums though: email, twitter, discord, etc ...

**Comment by danmosemsft** [comment]
cc @eerhardt since we were just talking about how this could be more approachable.

**Comment by eerhardt** [comment]
> My personal goal is to make it possible to use dotnet build and dotnet run for .cs files without needing an explicit project file.

@jaredpar - for other languages that allow this, the source files allow for references to external code inside of the source file (both to other files and packages). How do you plan on tackling this problem? The scenario works great for "Hello World", but as soon as you need to use anything that isn't in `Microsoft.NETCore.App`, or to include some other class file, now you need a project file.

**Comment by jaredpar** [comment]
@eerhardt  

By borrowing from existing directive style solutions 😄 

```cs
#!r Newtonsoft.Json 
...
```

Directives are very easy to scan and easy to ignore by the compiler. This can be parsed out with some `ReadLine` calls and a simple string matching code. That can be done by the layer of the build that creates the implicit project file and it can translate them to `<PackageReference>` or file references.

**Comment by km-hussain-in** [comment]
Copy the cscc script from https://github.com/km-hussain-in/CSharpCore/blob/master/cscc.zip into dotnet installation directory and use cscc command. It compiles .cs file to exe which you can run with dotnet command

**Comment by kinglionsz** [comment]
> Copy the cscc script from https://github.com/km-hussain-in/CSharpCore/blob/master/cscc.zip into dotnet installation directory and use cscc command. It compiles .cs file to exe which you can run with dotnet command

Good!

Tks so much👍👏

**Comment by km-hussain-in** [comment]
It works for me on Windows as well as Linux.

**Comment by km-hussain-in** [comment]
I have updated the scripts to work from any directory provided dotnet command is in the path

**Comment by mirabilos** [comment]
@km-hussain-in it doesn’t work on Debian; running with `bash -x` shows:

```
[…]
+ dotnet /usr/bin/sdk/5.0.201/Roslyn/bincore/csc.dll -nologo @/tmp/csc-5.0.4.rsp greeter.cs -t:library
Could not execute because the specified command or file was not found.
Possible reasons for this include:
  * You misspelled a built-in dotnet command.
  * You intended to execute a .NET program, but dotnet-/usr/bin/sdk/5.0.201/Roslyn/bincore/csc.dll does not exist.
  * You intended to run a global tool, but a dotnet-prefixed executable with this name could not be found on the PATH.
```

**Comment by evandrix** [comment]
> Download https://github.com/km-hussain-in/CSharpCore/blob/master/cscc.zip and follow the instructions in readme.txt.

is HTTP 404 ... someone re-up?

**Comment by mirabilos** [comment]
evandrix dixit:

>is HTTP 404 ... someone re-up?

Not exactly the same, but I use this:

	$ sh csc -nologo NinetyNineBottles.cs
	$ sh dotnet-mkrtc NinetyNineBottles.exe
	$ dotnet NinetyNineBottles.exe
	99 bottles of beer on the wall, 99 bottles of beer.
	Take one down and pass it around, 99 bottles of beer on the wall.

	98 bottles of beer on the wall, 98 bottles of beer.
	[…]

I had to split out the .runtimeconfig.json generation because I also
use csc for .dll generation, which don’t need it.

Enjoy,
//mirabilos
-- 
FWIW, I'm quite impressed with mksh interactively. I thought it was much
*much* more bare bones. But it turns out it beats the living hell out of
ksh93 in that respect. I'd even consider it for my daily use if I hadn't
wasted half my life on my zsh setup. :-) -- Frank Terbeck in #!/bin/mksh
#!/bin/sh

DOTNET_CLI_TELEMETRY_OPTOUT=1
export DOTNET_CLI_TELEMETRY_OPTOUT

sdkver=$(LC_ALL=C dotnet --version)
fwkver=$(LC_ALL=C dotnet --list-runtimes | \
    LC_ALL=C sed --posix -n '/^Microsoft.NETCore.App \([^ ]*\) .*$/{s//\1/p;q;}')

# dotnet-sdk-5.0 installed via .deb package
dotnethome=/usr/share/dotnet
dotnetlib=$dotnethome/shared/Microsoft.NETCore.App/$fwkver
dotnet_cscdll=$dotnethome/sdk/$sdkver/Roslyn/bincore/csc.dll
dotnet_csclib='-r:netstandard.dll -r:Microsoft.CSharp.dll -r:System.dll'
for x in "$dotnetlib"/System.*.dll; do
	dotnet_csclib="$dotnet_csclib -r:${x##*/}"
done
# add if needed
#dotnet_csclib="$dotnet_csclib -r:Microsoft.Win32.Primitives.dll"

exec dotnet "$dotnet_cscdll" "-lib:$dotnetlib" $dotnet_csclib "$@"

#!/bin/sh

DOTNET_CLI_TELEMETRY_OPTOUT=1
export DOTNET_CLI_TELEMETRY_OPTOUT

sdkver=$(LC_ALL=C dotnet --version)
fwkver=$(LC_ALL=C dotnet --list-runtimes | \
    LC_ALL=C sed --posix -n '/^Microsoft.NETCore.App \([^ ]*\) .*$/{s//\1/p;q;}')

exename=$1
case $exename in
(*.exe|*.EXE) ;;
(*)
	echo >&2 "E: $exename is not a .exe file"
	exit 1
	;;
esac

jsonname=${exename%.*}.runtimeconfig.json
printf '%s"%s"%s\n' \
    '{"runtimeOptions":{"framework":{"name":"Microsoft.NETCore.App","version":' \
    "$fwkver" '}}}' >"$jsonname"

**Comment by joshcangit** [comment]
Thanks to @mirabilos, I made a script to use as `csc` but I think `dotnet csc` is better and more convenient.

**Comment by km-hussain-in** [comment]
reuploaded dotnet-csc 
https://github.com/km-hussain-in/CourseSamples/blob/main/CSharp/cscc.zip?raw=true

**Comment by jhudsoncedaron** [comment]
I've come back to this now. The problem we are having is msbuild is collapsing under its own weight and we are actively considering if a makefile-like build solution would *actually be better*.

Issues:

https://github.com/dotnet/msbuild/issues/9179
https://github.com/dotnet/sdk/issues/38575
https://github.com/dotnet/aspnetcore/issues/58948
https://stackoverflow.com/questions/79415852/deterministictrue-deterministic-and-file-build-order (problem was found to be in msbuild not csc)
https://github.com/dotnet/msbuild/issues/10795 (only workaround is to stop using *; long discussion at https://stackoverflow.com/questions/77929595/how-do-i-late-resolve-in-csproj-files)

The general problem is all too clear; the fundamental architecture of msbuild is wrong and needs to change but changing it would break lots of people's csproj files; we need to raise up a completely different solution in parallel and leave msbuild in place while that is done. The only way to do *that* is ultimately to be able to invoke the compiler directly from the command line.

**Comment by Grizzlly** [comment]
@jaredpar Hello! Any plans on being able to invoke the compiler directly?

**Comment by baronfel** [comment]
No, there are no plans to do that. .NET projects today are _much more_ than just the compiler. Analyzers, dependencies, post-compilation packaging steps, etc. Anyone manually invoking the compiler is _very likely_ doing it wrong and at best setting themselves up for subtle mismatches and incompatibilities down the line.

**Comment by jhudsoncedaron** [comment]
@baronfel : Then you better start listening to complaints about problems with msbuild.

**Comment by baronfel** [comment]
I certainly hear you there. We're always investing in making MSBuild more performant and fixing problems like some of the ones you've linked. I can't help but notice that most of your issues are tied to the relationship between solution files and MSBuild - I hate this too. Solution files are fundamentally a concept that MSBuild doesn't control and that are tightly tied to Visual Studio and the way that IDE processes things - so to some degree our hands are tied there. Changing things at that level is much more than just updating the SDK or updating MSBuild - there's _tons_ of cross-team work that has to be done to ensure tooling experiences are also made aware of any changes we make, and that's much harder to get coordinated and make any progress on.

In the meantime, we're working on what we _can_ control - we're making MSBuild faster through supporting multithreaded, single-process builds, we're making it easier to work with for MSBuild logic authors by letting you use .NET-only Tasks from .NET Framework MSBuild (aka VS), and we're working on decoupling VS and the .NET SDK's MSBuild so that we can iterate faster through the SDK without having to be perfectly in sync with VS.

**Comment by jhudsoncedaron** [comment]
@baronfel : Yeah; most of them are, but there's a couple that aren't.

`dotnet pack` has been making broken nuget packages for most of a decade, but that's probably unrelated.

Related: I noticed that VS and VS Code have an acceptable amount of intellisene when you use open directory on a directory with a project file that can't parse due to a typo.

**Comment by baronfel** [comment]
> `dotnet pack` has been making broken nuget packages for most of a decade, but that's probably unrelated.

Say more/link to an issue going into more detail please?

**Comment by jhudsoncedaron** [comment]
@baronfel : Here's one. https://github.com/NuGet/Home/issues/13853

**Comment by baronfel** [comment]
You're right - that is unrelated. I do agree that actual content-types should be used, for what that's worth, but you've raised it in the right place. If it bothers you,  I'd suggest sending a PR to see if any tests break - Jeff has linked to the location that would need to be updated.

**Comment by jhudsoncedaron** [comment]
@baronfel : The reason I didn't do that is it's insufficient by itself. I would somehow have to get a PR approved that fixes all of them, where if any one part fails there's no point merging any of them.

**Comment by mirabilos** [comment]
Manually invoking the compiler through a Makefile is *not* wrong, and wholly sufficient for a two-source-file library.

**Comment by mirabilos** [comment]
Also, Makefiles are the only way to compile .NET things for which there is sufficient documentation and accessibility.

**Comment by baronfel** [comment]
I'm going to close this issue - we are not going to support this, and we're not going to encourage folks to go color outside the lines here. The support costs on the team are too big to even entertain the thought.

**Comment by mirabilos** [comment]
> The support costs

You’re **Microsoft,** how much bigger would you need to get to offer this. Honestly. And here I’m a FOSS hobbyist and support running my software on Minix-386, DEC ULTRIX and even 1991’s Xenix…

But, okay. I did touch this for a work project only, and I *do* have a working, if ugly and fragile, solution. Sharing <https://github.com/qvest-digital/ECN-Bits/tree/master/dotnet> here for anyone else coming this way. Cheers!

**Comment by jhudsoncedaron** [comment]
The support cost of csproj is eating us alive here.

**Comment by KalleOlaviNiemitalo** [comment]
If the request is not going to be implemented then it would be nicer to close as "Not planned" rather than as "Completed".

**Comment by mirabilos** [comment]
But @KalleOlaviNiemitalo to change that now would cost that poor tiny company soooo much! 😹

**Comment by TheAjaykrishnanR** [comment]
I have made a tool that aot compiles single `.cs` files into native executables, if anybody is interested:

https://github.com/TheAjaykrishnanR/dflat

Supports .NET 10
